### ISSUE TITLE
Unable to deploy planner to prod preview

### ISSUE THREAD

**Issue by nimishamukherjee** [openshiftio/openshift.io#2749]
https://github.com/fabric8io/fabric8-build-team/issues/48

The planner release step is failing.

Pipeline error
https://jenkins.cd.test.fabric8.io/blue/organizations/jenkins/fabric8-ui%2Ffabric8-planner/detail/master/201/pipeline

**Comment by michaelkleinhenz** [comment]
This is SEV-1 as there is no workaround and we're seriously blocked for next weeks demo.

**Comment by aslakknutsen** [comment]
Similar https://jenkins.cd.test.fabric8.io/blue/organizations/jenkins/fabric8-services%2Ffabric8-tenant/detail/master/218/pipeline

**Comment by joshuawilson** [comment]
This is issue with the file system.
`chmod: changing permissions of ‘/root/.ssh-git/ssh-key’: Read-only file system`

The next command should be:
`chmod 600 /root/.ssh-git/ssh-key.pub`

But it doesn't get that far. It has to get that far before it can even get to the release step. So it isn't the release that is failing.

**Comment by jarifibrahim** [comment]
I don't understand why do we even need the `ssh-key` and `ssh-key.pub`. We have the Github token available at https://github.com/fabric8io/fabric8-pipeline-library/blob/master/vars/npmRelease.groovy#L21.
The token alone should be enough for all authentication/validation purposes.

**Comment by joshuawilson** [comment]
need to talk to Build team. @rupalibehera @chmouel @pradeepto

**Comment by jaseemabid** [comment]
I've been looking at this since morning and I've some updates. 

This is indeed an issue with the secrets mounted read-only and us doing a chmod on it. @sthaha and I confirmed it today morning. 

The solution is to change the `defaultMode` for the volume's secret in the Pod config yaml as mentioned in https://kubernetes.io/docs/concepts/configuration/secret such that we don't have to chmod after the mount. Setting the umask seems like the right approach.  

I could not find the Pod definition for the Jenkins slave to change this config and asked @surajssd and Shubham for help and they couldn't find it either. I'm not sure if there is more config for the Jenkins slave than what is done with `podTemplate`s in pipeline library and we will have to wait for @rupalibehera to come back and give us more insight into how this is wired together. Once we find out how the Pod template yaml is generated, we can fix it with this patch. 

tldr; Too many levels of abstraction and indirection, cannot find out where the configuration actually happens.

**Comment by joshuawilson** [comment]
Thanks for digging into this. Please keep us posted as you learn more. 

This is broken in more then Planner.

**Comment by chmouel** [comment]
So digging a bit more about this and document how things works : 

* We have a secret called `jenkins-git-ssh`

* The secret get mounted by Jenkins when starting a new pod with this snippet in its xml configuration : 
```java
                    <org.csanchez.jenkins.plugins.kubernetes.volumes.SecretVolume>
                          <mountPath>/root/.ssh-git</mountPath>
                          <secretName>jenkins-git-ssh</secretName>
                        </org.csanchez.jenkins.plugins.kubernetes.volumes.SecretVolume>
```
* Now at startup `fabric8-pipeline-library` try to fix the user permissions in a lot of places : 

![image](https://user-images.githubusercontent.com/98980/37976539-02e5f7d6-31e2-11e8-93fe-33d50308594f.png)

* The proper way should be to setup the defaultMode bits on the kubernetes volume with `defaultMode: ` part on the volume template

* Unfortunately and this is where things get complicated for me jenkins uses this code to mount the secret as volume : 

https://github.com/fabric8io/kubernetes-plugin//blob/a4e8ecb/src/main/java/org/csanchez/jenkins/plugins/kubernetes/volumes/SecretVolume.java#L50

* This is calling on `[kubernetes-model](https://github.com/fabric8io/kubernetes-model)`generated models to talk too kubernetes API

* This don't seem to support `defaultMode` (but I may be wrong, it's hard to grep in auto generated code)

## To be done

I assume the proper way would be : (need to be done by some people who have some java skills ) : 

* [ ] Get kubernetes-models volumeBuild to support `defaultMode` 
* [ ] Give to kubernetes-plugin's                 `org.csanchez.jenkins.plugins.kubernetes.volumes.SecretVolume` an argument which accept an optional `defaultMode` and pass it to the `volumeBuild`
* [ ] Get this argument onto jenkins config.xml stored into the configmap called `jenkins` in GKE 
* [ ] Patch `fabric8-pipeline-library` to not attempt to chmod/chown the ssh keys when on a read-only mountPath `[[ -w /root/.ssh-git ]] || return`

The hacky alternative would be to do this in fabric8-pipeline-library  : 

* [ ] mount the secret to /root/.ssh-git-ro
* [ ] change jenkins-pipeline-library to  do this at first : 
```bash
[[ -d /root/.ssh-git && -e /root/.ssh-git-ro ]] || { 
cp -a /root/.ssh-git-ro/ /root/.ssh-git
}
````

Theorically that should work, may need some digging, 

@jaseemabid can you try the hacky solution and see if that works tomorrow ?
@pradeepto can we get a java expert working on the proper solution if we have resources for that?

**Comment by jaseemabid** [comment]
@chmouel I'm actually wondering how these keys are used at all. As far as I know, you need an `ssh-add` on the key file after starting the agent and I cannot find that code anywhere in pipeline library. Are we doing some token setup somewhere? Will it be ok if we just delete these 4 lines? I'll give that a try tomorrow.

**Comment by joshuawilson** [comment]
Since this was working, do we know what changed to make it not work anymore?

**Comment by chmouel** [comment]
@joshuawilson  https://github.com/kubernetes/kubernetes/issues/60814

**Comment by chmouel** [comment]
@jaseemabid sounds like a good idea 👍, i could not find a reference to it as well but there is so many pieces there that it's hard to figure out what uses what,

**Comment by jaseemabid** [comment]
Fixed at https://github.com/fabric8io/fabric8-pipeline-library/commit/4b4943461f4f90987d166ae79c0b3d7c7f11babb

**Comment by joshuawilson** [comment]
@jaseemabid your commit calls it a "temporary workaround", do you have a permanent solution in the works?

**Comment by joshuawilson** [comment]
I'm leaving this open but downgrading the critical nature of it since a work around is in place.

**Comment by jarifibrahim** [comment]
@jaseemabid The issue still exists. Check this build https://jenkins.cd.test.fabric8.io/job/fabric8-ui/job/fabric8-npm-dependencies/job/master/178/console

**Comment by jaseemabid** [comment]
@joshuawilson Yep. I'll clean up pipeline library later, but that shouldn't affect the consumers of the library.

**Comment by nimishamukherjee** [comment]
@joshuawilson @jaseemabid @pradeepto closing this issue as Planner's release issue has been fixed.
Have created #2825 to track the issue around failure of auto update palnner's version to fabric8-ui.
